### PR TITLE
bug 1238587 - Python 3.4 -> 3.5, pip constraints and hashes 

### PR DESCRIPTION
**Pull request by jwhitlock**
Make some changes to better support Python 3 and 3.5 specifically:

* Use print as a function in integration tests,
* Don't mutate a dictionary during iteration
* Read JSON into an OrderedDict to preserve declaration order and hit all the error-handling code branches
* Use Python 3.5 as "base python" in TravisCI

Distribute ``requirements.txt`` and ``requirements-dev.txt`` across files in new ``requirements`` folder:
* ``requirements/application.txt`` includes the minimum requirements to run a local server
* ``requirements/production.txt`` adds decencies to run in the Heroku environment
* ``requirements.txt`` just includes ``requirements/production.txt``
* tox targets now load specific requirements files rather than also specify versions (for example, ``tox flake8`` uses ``requirements/flake8.txt``)
* ``requirements/development.txt`` loads everything

Finally, use some new bits of pip:
* pip 7.x added constraints (specified in ``requirements/constraints.txt``), which specify package versions if a requirement needs them, but won't install a package if nothing asks for it. This is perfect for packages like ``django-allauth``, which load different packages for open ID support in Python 2 vs Python 3.  It also means we stop installing packages if a required package stops requiring it.
* pip 8.x added packages hashes, used to verify packages downloaded from PyPI, to mitigate man-in-the-middle attacks and erroneous package caches. This requires fully specifying any dependencies, so is a good match for constraints.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,13 @@
 # Config file for automatic testing at travis-ci.org
 sudo: false
 language: python
+python: "3.5"
```
Comment (willkg) [inline .travis.yml]: Hooray!

File: MANIFEST.in [hunk 2]
```diff
@@ -16,6 +16,7 @@ recursive-include tools *.py *.sh
 recursive-include mdn *.py
 recursive-include mdn/jinja2 *.html
 recursive-include mdn/static *.css
+recursive-include requirements *.txt
```
Comment (willkg) [inline MANIFEST.in]: I don't see a `requirements/` directory. What does this line pick up?

Comment (willkg) [inline MANIFEST.in]: My bad. I see it now. I hate github's interface for viewing things sometimes.

Comment (willkg) [inline MANIFEST.in]: Do we also need to pick up `requirements-rtd.txt`? Should that file be in `requirements`, too?

File: tox.ini [hunk 3]
```diff
@@ -18,33 +18,27 @@ deps = {[base]deps}
...
 deps =
     {[base]deps}
-    coverage==4.0.3
-    coveralls==1.1
+    -r{toxinidir}/requirements/coverage.txt
 commands =
     coverage run --source bcauth,mdn,webplatformcompat,tools setup.py test
     coverage report -m
-    coveralls
+    -coveralls
```
Comment (willkg) [inline tox.ini]: What does the `-` sign here do? Does that mean "run this thing, but if it dies, don't consider that an error" or something like that?

**Comment by willkg** [comment]
This looks fine to me. It's super benign despite having a bazillion commits and touching a bunch of files.

I kind of think the way the requirements files are organized is more work than I would have done, but I may not really understand the issues.

r+

File: tox.ini (hunk 3 above)
Comment (jwhitlock) [inline tox.ini]: Yes, if ``coveralls`` returns a non-zero exit code (which it does locally), then ``tox coverage`` is still a success.

**Comment by jwhitlock** [comment]
I agree, the requirements organization is probably overkill. One big requirements file might be the way to go (and is closer to the 12factor philosophy). If the next package update is too difficult, I may go that route.

The last commit added was to add more hashes from [hashin](http://www.peterbe.com/plog/bestest-python-dependencies). I'll need to force-push to fix it.

File: MANIFEST.in (hunk 2 above)
Comment (jwhitlock) [inline MANIFEST.in]: I may need to move it to /requirements, yeah. It is for readthedocs, which installs it's own preferred version of Sphinx, etc.  If it tries to install a package with a C-dependency (like psycopg), then the docs don't build.  I'll probably have to do a two-phase move - add an empty file in this PR, update the RTD config, remove requirements-rtd.txt in the next PR.

**Comment by jwhitlock** [comment]
Force push to (hopefully) fix TravisCI (install latest pip and wheel), and additional commit to move [readthedocs](https://readthedocs.org) config file into the new place.

**Comment by jwhitlock** [comment]
Force-push to add hashes for ``python-openid``

**Pull request merged**